### PR TITLE
Message tracing support

### PR DESCRIPTION
**Pull request by AnujaLK**

### CODE REVIEW

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/inbound/MessagePreProcessor.java [hunk 1]
```diff
@@ -268,6 +269,10 @@ private AndesMessage cloneAndesMessageMetadataAndContent(AndesMessage message) {
      */
     private void setMessageID(AndesMessage message) {
         long messageId = idGenerator.getNextId();
+
+        //Tracing message
+        MessageTracer.trace(message, MessageTracer.MESSAGE_ID_MAPPED + " id: " + messageId);
```
Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/inbound/MessagePreProcessor.java]: can't we do like : isTracingEnabled then do this?
 this is in the performance path...

File: modules/andes-core/broker/src/main/java/org/wso2/andes/tools/utils/MessageTracer.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    public static final String SLOT_INFO_UPDATED = "slot information updated";
+    public static final String METADATA_READ_FROM_DB = "metadata read from database";
+    public static final String METADATA_BUFFERED_FOR_DELIVERY = "metadata buffered for delivery";
+    public static final String CONTENT_READ = "content read from database";
+    public static final String DISPATCHED_TO_PROTOCOL ="dispatched to protocol level for delivery";
+
+
+
+    public static void trace(AndesMessage message, String content) {
+        if(log.isDebugEnabled()) {
```
Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/tools/utils/MessageTracer.java]: loggers support Trace level logs isn't it? why not we use that? than debug.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/inbound/MessagePreProcessor.java (hunk 1 above)
Comment (AnujaLK) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/inbound/MessagePreProcessor.java]: This has been performed inside MessageTracer class.

Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/inbound/MessagePreProcessor.java]: nope thats not the point. point is we do the concatenation :
MessageTracer.MESSAGE_ID_MAPPED + " id: " + messageId

for each inbound event. which is not something we want to do  unless tracing is enabled.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/inbound/StateEventHandler.java [hunk 3]
```diff
@@ -107,6 +108,11 @@ public void updateSlotsAndQueueCounts(InboundEventContainer eventContainer) {
             // wise.
             messagingEngine.incrementQueueCount(message.getMetadata().getDestination(), 1);
 
+            //Tracing Message
+            MessageTracer.trace(message, MessageTracer.SLOT_INFO_UPDATED);
+
+            eventContainer.pubAckHandler.ack(message.getMetadata());
```
Comment (pumudu88) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/inbound/StateEventHandler.java]: This pub ack handler is not relevant here. removed in PR https://github.com/wso2/andes/pull/192

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/inbound/MessagePreProcessor.java [hunk 4]
```diff
@@ -268,6 +269,12 @@ private AndesMessage cloneAndesMessageMetadataAndContent(AndesMessage message) {
      */
     private void setMessageID(AndesMessage message) {
         long messageId = idGenerator.getNextId();
+
+        //Tracing message
+        if(MessageTracer.isEnabled()) {
```
Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/inbound/MessagePreProcessor.java]: wrong indentation?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/tools/utils/MessageTracer.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.andes.tools.utils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
```
Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/tools/utils/MessageTracer.java]: wrong imports or what?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/inbound/MessagePreProcessor.java (hunk 4 above)
Comment (AnujaLK) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/inbound/MessagePreProcessor.java]: This is already fixed with another PR.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/tools/utils/MessageTracer.java (hunk 5 above)
Comment (AnujaLK) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/tools/utils/MessageTracer.java]: Fixed with PR https://github.com/wso2/andes/pull/276/files

File: modules/andes-core/broker/src/main/java/org/wso2/andes/tools/utils/MessageTracer.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    public static final String DISPATCHED_TO_PROTOCOL ="dispatched to protocol level for delivery";
+
+	/**
+	 * This method will print debug logs for message activities. This will accept andes message as
+	 * a parameter
+	 * @param message Andesmessage
+	 * @param content Message activity
+	 */
+    public static void trace(AndesMessage message, String content) {
+        if(log.isTraceEnabled()) {
```
Comment (AnujaLK) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/tools/utils/MessageTracer.java]: Indentation already fixed.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/tools/utils/MessageTracer.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+	/**
+	 * This method will print debug logs for message activities. This will accept metadata as
+	 * a parameter
+	 * @param metadata andes metadata object
+	 * @param content message activity
+	 */
+    public static void trace(AndesMessageMetadata metadata, String content) {
+        if(log.isTraceEnabled()) {
```
Comment (AnujaLK) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/tools/utils/MessageTracer.java]: Indentation already fixed.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/tools/utils/MessageTracer.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+	/**
+	 * This method will print debug logs for message activities. This will accept andes message as
+	 * a parameter
+	 * @param message Andesmessage
+	 * @param content Message activity
+	 */
+    public static void trace(AndesMessage message, String content) {
+        if(log.isTraceEnabled()) {
+            String messageContent = "Message {";
+            messageContent += " Destination: " + message.getMetadata().getDestination();
```
Comment (hemikak) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/tools/utils/MessageTracer.java]: We discourage to use += sign.

Comment (AnujaLK) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/tools/utils/MessageTracer.java]: This is already replaced by StringBuilder in another PR.

**Pull request merged**